### ISSUE TITLE
wgengine/magicsock: DERP reading broken from refactor?

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#1282]
hello has been having problems lately (#1266, etc). Its goroutines include:
```
goroutine 199 [select, 134 minutes]:
tailscale.com/wgengine/magicsock.(*Conn).runDerpReader(0xc0000d4000, 0xd3ee20, 0xc0029840c0, 0x0, 0xffff7f030328, 0xc00000c100, 0x4, 0xc0020e8180, 0xc003afe150, 0xc000182600)
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/magicsock/magicsock.go:1451 +0xb56
created by tailscale.com/wgengine/magicsock.(*Conn).derpWriteChanOfAddr
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/magicsock/magicsock.go:1300 +0xb05
```

Where that 134 minute select is here:

```go
                // Before we wake up ReceiveIPv4 with SetReadDeadline,                                                          
                // note that a DERP packet has arrived.  ReceiveIPv4                                                            
                // will read this field to note that its UDP read                                                               
                // error is due to us.                                                                                          
                atomic.AddInt64(&c.derpRecvCountAtomic, 1)
                // Cancel the pconn read goroutine.                                                                             
                c.pconn4.SetReadDeadline(aLongTimeAgo)

                select {
                case <-ctx.Done():
                        return
                case c.derpRecvCh <- res:
                        select {
                        case <-ctx.Done():
                                return
                        case <-didCopy:
                                continue
                        }
                }
```

The `receiveIPv4DERP` method is the only thing that reads from that channel, and it's only called:

```go
func (c *Conn) ReceiveIPv4(b []byte) (n int, ep conn.Endpoint, err error) {
        for {
                n, pAddr, err := c.pconn4.ReadFrom(b)
                if err != nil {
                        // If the pconn4 read failed, the likely reason is a DERP reader received
                        // a packet and interrupted us.
                        // It's possible for ReadFrom to return a non deadline exceeded error
                        // and for there to have also had a DERP packet arrive, but that's fine:
                        // we'll get the same error from ReadFrom later.
                        if c.derpPacketArrived() {
                                c.pconn4.SetReadDeadline(time.Time{}) // restore
                                n, ep, err = c.receiveIPv4DERP(b)
                                if err == errLoopAgain {
                                        continue
                                }
                                return n, ep, err
                        }
                        return 0, nil, err
                }
```

Where:

```go
func (c *Conn) derpPacketArrived() bool {
        rc := atomic.LoadInt64(&c.derpRecvCountAtomic)
        if rc != c.derpRecvCountLast {
                c.derpRecvCountLast = rc
                return true
        }
        return false
}
```

This was all added in the 5c619882 refactor (for performance, sigh).

But it seems that there's a bug in there.

Further, that commit was included in 1.4.0.

I have some ideas, but might not be able to look energetically enough for a few days.

/cc @danderson @crawshaw @josharian (any eyeballs welcome)

<img src="https://frontapp.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_2rpxt)

**Comment by bradfitz** [comment]
Oh! The new scheme in 1.4 isn't correct when there are two concurrent DERP reads coming in (from different DERPs). They can both wake up the UDP reader and one can get stuck forever.

**Comment by bradfitz** [comment]
Okay, this is all very clear now. When I did the DERP->ReceiveIPv4 wake up mechanism I straight up forgot there were N DERP readers all injecting into 1 UDP reader. The mechanism from https://github.com/tailscale/tailscale/commit/5c619882bc4911a2c9e7d0bb491b9e50d27afcd7 only works 1:1.

I just pushed a new test https://github.com/tailscale/tailscale/commit/6d2b8df06df1d20d4ae92793a066340bb26b6e25 that super easily reproduces the hang. The new test currently only uses 1 DERP reader goroutine for now, until this bug is fixed. I put a temporary flag in the code if you want to set `--test-issue-1282` and see it hang, but it hangs exactly as you'd expect.

Now that the test is in, I'll do the fix + backport.
